src/main/java/io/pravega/perf/WriterWorker.java

```diff
@@ -57,67 +74,116 @@ public Void call() throws InterruptedException, ExecutionException, IOException
-                String key;
-                if (isRandomKey) {
-                    key = Integer.toString(workerID + rand.nextInt());
-                } else {
-                    key = Integer.toString(workerID);
-                }
 
-                final Instant startTime = Instant.now();
-                retFuture = writeData(key, payload);
-                // event ingestion
-                retFuture = stats.recordTime(retFuture, startTime, payload.length());
-                stats.print();
-                tput.control(stats.eventsRate());
+    private class EventsWriterTime implements Performance {
 
-            }
+        public void benchmark() throws InterruptedException,  IOException {
+            final long msToRun = secondsToRun * MS_PER_SEC;
+            long time = System.currentTimeMillis();
 
+            for (int i = 0; (time - StartTime) < msToRun; i++) {
+                time = recordWrite(payload, stats::recordTime);
+                eCnt.control(i);
+            }
             flush();
+        }
+    }
+
+    private class EventsWriterTimeRW implements Performance {
 
-            //Wait for the last packet to get acked
-            retFuture.get();
+        public void benchmark() throws InterruptedException, IOException {
+            final long msToRun = secondsToRun * MS_PER_SEC;
+            long time = System.currentTimeMillis();
+
+            for (int i = 0; (time - StartTime) < msToRun; i++) {
+                time = System.currentTimeMillis();
+                final String val = time + ", " + workerID + ", ";
+                final String data = (val + payload).substring(0, messageSize);
+                writeData(data);
+                eCnt.control(i);
+            }
+            flush();
         }
     }
 
-    private class ThroughputWriter implements Performance {
-
-        public void benchmark() throws InterruptedException, ExecutionException, IOException {
-            CompletableFuture retFuture = null;
-            Random rand = new Random();
-
-            while (Duration.between(StartTime, Instant.now()).getSeconds() < secondsToRun) {
-                // Construct event payload
-                String val = System.currentTimeMillis() + ", " + workerID + ", " + (int) (Math.random() * 200);
-                String payload = String.format("%-" + messageSize + "s", val);
-                String key;
-                if (isRandomKey) {
-                    key = Integer.toString(workerID + rand.nextInt());
-                } else {
-                    key = Integer.toString(workerID);
-                }
+    private class EventsController {
+        private static final long NS_PER_MS = 1000000L;
+        private static final long NS_PER_SEC = 1000 * NS_PER_MS;
+        private static final long MIN_SLEEP_NS = 2 * NS_PER_MS;
+        private final long startTime;
+        private final long sleepTimeNs;
+        private final int eventsPerSec;
+        private long toSleepNs  = 0;
```

Comment:
What's going on here? This is clearly not final, it's being modified from a public method, There is no annotation declaring where the lock is. How is anyone supposed to verify if this is ok or not?

Reply:
now, I have tagged EventsController as  NotThreadSafe;
however, each producer will have its own EventsController  object, so no multiple threads/writers accessing the single EventsController  .